File: 프로그래머스/2/131704. 택배상자/택배상자.cpp
#include <string>
#include <vector>
#include <stack>
using namespace std;

int solution(vector<int> order) {
    int answer = 0;
    vector<int> mainBelt (order.size());
    for(int i = 0; i < order.size(); ++i)
    {
        mainBelt[order[i]-1] = i;
    }
    
    stack<int> subBelt;
    
    int target = 0;
    for(int i = 0; i < mainBelt.size(); ++i)
    {
        subBelt.push(mainBelt[i]);
        
        while(!subBelt.empty() && subBelt.top() == target)
        {
            subBelt.pop();
            ++target;
            ++answer;
        }
    }
    
    
    return answer;
}

File: 프로그래머스/2/131704. 택배상자/택배상자.java
import java.util.*;

class Solution {
    public int solution(int[] order) {
        int answer = 0;
        int[] mainBelt = new int[order.length];
        for (int i = 0; i < order.length; i++) {
            mainBelt[i] = i + 1;
        }

        int orderIdx = 0;
        Deque<Integer> stack = new ArrayDeque<>();

        for (int i = 0; i < mainBelt.length; i++) {
            stack.push(mainBelt[i]);
            
            while (!stack.isEmpty() && order[orderIdx] == stack.peek()) {
                stack.pop();
                orderIdx++;
                answer++;
            }
        }

        return answer;
    }
}